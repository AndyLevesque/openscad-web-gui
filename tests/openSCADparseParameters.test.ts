import fs from 'fs';

import parseParameters from '../src/lib/openSCAD/parseParameter';

describe('testing parameter parsing of openscad scripts', () => {
  const testFixture = (path) => (fileName) => {
    if (fileName.endsWith('.scad')) {
      test(`testing ${fileName}`, () => {
        const script = fs.readFileSync(path + fileName, 'utf8');
        const parameters = parseParameters(script);
        expect(parameters).toMatchSnapshot(fileName);
      });
    }
  };

  const pathWithTestFixtures = [
    __dirname + '/fixtures/openSCAD/customizer/',
    __dirname + '/fixtures/openSCAD/',
  ];

  pathWithTestFixtures.forEach((path) => {
    fs.readdirSync(path).forEach(testFixture(path));
  });
});
