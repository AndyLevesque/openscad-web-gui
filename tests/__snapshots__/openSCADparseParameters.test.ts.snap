// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing parameter parsing of openscad scripts testing checkbox.scad: checkbox.scad 1`] = `
[
  {
    "description": "description",
    "group": undefined,
    "name": "Variable",
    "type": "boolean",
    "value": true,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing dropDownBox.scad: dropDownBox.scad 1`] = `
[
  {
    "description": "combo box for number",
    "group": undefined,
    "name": "Numbers",
    "type": "number",
    "value": 2,
  },
  {
    "description": "combo box for string",
    "group": undefined,
    "name": "Strings",
    "type": "string",
    "value": "foo",
  },
  {
    "description": "labeled combo box for numbers",
    "group": undefined,
    "name": "Labeled_values",
    "type": "number",
    "value": 10,
  },
  {
    "description": "labeled combo box for string",
    "group": undefined,
    "name": "Labeled_value",
    "type": "string",
    "value": "S",
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing example1.scad: example1.scad 1`] = `
[
  {
    "description": "combo box for number",
    "group": "Drop down box:",
    "name": "Numbers",
    "type": "number",
    "value": 2,
  },
  {
    "description": "combo box for string",
    "group": "Drop down box:",
    "name": "Strings",
    "type": "string",
    "value": "foo",
  },
  {
    "description": "labeled combo box for numbers",
    "group": "Drop down box:",
    "name": "Labeled_values",
    "type": "number",
    "value": 10,
  },
  {
    "description": "labeled combo box for string",
    "group": "Drop down box:",
    "name": "Labeled_value",
    "type": "string",
    "value": "S",
  },
  {
    "description": "slider widget for number",
    "group": "Slider",
    "name": "slider",
    "type": "number",
    "value": 34,
  },
  {
    "description": "step slider for number",
    "group": "Slider",
    "name": "stepSlider",
    "type": "number",
    "value": 2,
  },
  {
    "description": "description",
    "group": "Checkbox",
    "name": "Variable",
    "type": "boolean",
    "value": true,
  },
  {
    "description": "spinbox with step size 1",
    "group": "Spinbox",
    "name": "Spinbox",
    "type": "number",
    "value": 5,
  },
  {
    "description": "Text box for vector with more than 4 elements",
    "group": "Textbox",
    "name": "Vector6",
    "type": "string",
    "value": "[12,34,44,43,23,23]",
  },
  {
    "description": "Text box for string",
    "group": "Textbox",
    "name": "String",
    "type": "string",
    "value": "hello",
  },
  {
    "description": "Text box for vector with less than or equal to 4 elements",
    "group": "Special vector",
    "name": "Vector1",
    "type": "string",
    "value": "[12]",
  },
  {
    "description": undefined,
    "group": "Special vector",
    "name": "Vector2",
    "type": "string",
    "value": "[12,34]",
  },
  {
    "description": undefined,
    "group": "Special vector",
    "name": "Vector3",
    "type": "string",
    "value": "[12,34,46]",
  },
  {
    "description": undefined,
    "group": "Special vector",
    "name": "Vector4",
    "type": "string",
    "value": "[12,34,46,24]",
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing printables-513382.scad: printables-513382.scad 1`] = `
[
  {
    "description": "How high should each segment be?",
    "group": "Shape parameters",
    "name": "segment_heights",
    "type": "string",
    "value": "[35, 35, 35, 35, 35]",
  },
  {
    "description": "How wide should each segment be?",
    "group": "Shape parameters",
    "name": "segment_diameters",
    "type": "string",
    "value": "[140, 100, 120, 90, 110, 90]",
  },
  {
    "description": "The shape of the twig",
    "group": "Shape parameters",
    "name": "twig_shape",
    "type": "string",
    "value": "circle",
  },
  {
    "description": "How much should the twigs be twisted?",
    "group": "Shape parameters",
    "name": "twig_twist",
    "type": "string",
    "value": "[25, 25, 25, 0, 25]",
  },
  {
    "description": "How much should the twigs in the counter direction be twisted?",
    "group": "Shape parameters",
    "name": "twig_counter_twist",
    "type": "string",
    "value": "[25, 25, 25, 0, 25]",
  },
  {
    "description": "How thick should a twig be?",
    "group": "Shape parameters",
    "name": "twig_diameter",
    "type": "number",
    "value": 3,
  },
  {
    "description": "How many twigs should there be?",
    "group": "Shape parameters",
    "name": "twig_count",
    "type": "number",
    "value": 25,
  },
  {
    "description": "How many twigs should be in the other direction?",
    "group": "Shape parameters",
    "name": "twig_counter_count",
    "type": "number",
    "value": 25,
  },
  {
    "description": "The height of the top ring",
    "group": "Shape parameters",
    "name": "top_height",
    "type": "number",
    "value": 15,
  },
  {
    "description": "The height of the bottom ring",
    "group": "Shape parameters",
    "name": "bottom_height",
    "type": "number",
    "value": 5,
  },
  {
    "description": "How big is the thing you want to put in here?",
    "group": "Socket parameters",
    "name": "socket_diameter",
    "type": "number",
    "value": 81,
  },
  {
    "description": undefined,
    "group": "Socket parameters",
    "name": "socket_step_top_offset",
    "type": "number",
    "value": 2,
  },
  {
    "description": undefined,
    "group": "Socket parameters",
    "name": "socket_step_height",
    "type": "number",
    "value": 2.5,
  },
  {
    "description": undefined,
    "group": "Socket parameters",
    "name": "socket_step_diameter",
    "type": "number",
    "value": 87,
  },
  {
    "description": "Override the general settings and let the randomizer go wild!",
    "group": "Randomize Parameters",
    "name": "override_values",
    "type": "boolean",
    "value": false,
  },
  {
    "description": "The seed for the random number generator.",
    "group": "Randomize Parameters",
    "name": "seed",
    "type": "number",
    "value": 1.1,
  },
  {
    "description": "The amount of segments will be chosen between those two numbers.",
    "group": "Randomize Parameters",
    "name": "min_max_segment_count",
    "type": "string",
    "value": "[2, 5]",
  },
  {
    "description": "The height of each segment is between those two numbers.",
    "group": "Randomize Parameters",
    "name": "min_max_heights",
    "type": "string",
    "value": "[0, 180]",
  },
  {
    "description": "Min/max of all the diameters.",
    "group": "Randomize Parameters",
    "name": "min_max_diameters",
    "type": "string",
    "value": "[100, 180]",
  },
  {
    "description": "The amount of twig twist will be between those two numbers.",
    "group": "Randomize Parameters",
    "name": "min_max_twig_twist",
    "type": "string",
    "value": "[0, 180]",
  },
  {
    "description": "Min/max of the twig diameter.",
    "group": "Randomize Parameters",
    "name": "min_max_twig_diameter",
    "type": "string",
    "value": "[2, 20]",
  },
  {
    "description": "The twig count of both twig directions will be between those two numbers.",
    "group": "Randomize Parameters",
    "name": "min_max_twig_count",
    "type": "string",
    "value": "[5, 15]",
  },
  {
    "description": undefined,
    "group": "Quality settings",
    "name": "shell_fn",
    "type": "number",
    "value": 5,
  },
  {
    "description": undefined,
    "group": "Others",
    "name": "show_labels",
    "type": "boolean",
    "value": false,
  },
  {
    "description": undefined,
    "group": "Others",
    "name": "show_parameters",
    "type": "boolean",
    "value": false,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing printables-682650.scad: printables-682650.scad 1`] = `
[
  {
    "description": "Size of the panel on the x-axis",
    "group": "Basics",
    "name": "length",
    "type": "number",
    "value": 180,
  },
  {
    "description": "Size of the panel on the y-axis",
    "group": "Basics",
    "name": "depth",
    "type": "number",
    "value": 180,
  },
  {
    "description": "Should a backplate be used?",
    "group": "Basics",
    "name": "use_backplate",
    "type": "boolean",
    "value": false,
  },
  {
    "description": "Size of a single cube on the x-axis",
    "group": "Cube parameters",
    "name": "cube_length",
    "type": "number",
    "value": 20,
  },
  {
    "description": "Size of a single cube on the y-axis",
    "group": "Cube parameters",
    "name": "cube_depth",
    "type": "number",
    "value": 20,
  },
  {
    "description": "Minimum height of a single cube",
    "group": "Cube parameters",
    "name": "cube_min_height",
    "type": "number",
    "value": 5,
  },
  {
    "description": "Maximum height of a single cube",
    "group": "Cube parameters",
    "name": "cube_max_height",
    "type": "number",
    "value": 80,
  },
  {
    "description": "Wall thickness of cube",
    "group": "Cube parameters",
    "name": "cube_wall_thickness",
    "type": "number",
    "value": 1,
  },
  {
    "description": "Top perimeter thickness",
    "group": "Cube parameters",
    "name": "cube_top_thickness",
    "type": "number",
    "value": 2,
  },
  {
    "description": "Which seed to use? -1 generates a new layout with each generation",
    "group": "Cube parameters",
    "name": "seed_value",
    "type": "string",
    "value": "-1",
  },
  {
    "description": undefined,
    "group": "Hanging holes",
    "name": "use_hanging_holes",
    "type": "boolean",
    "value": true,
  },
  {
    "description": "The thickness of the baseplate where the opening for the screw/nail is",
    "group": "Hanging holes",
    "name": "wall_hole_thickness",
    "type": "number",
    "value": 1,
  },
  {
    "description": "The diameter of the center hole",
    "group": "Hanging holes",
    "name": "wall_hole_center_size",
    "type": "number",
    "value": 7,
  },
  {
    "description": "The diameter of the screw so it can slide along the hanging hole handles",
    "group": "Hanging holes",
    "name": "wall_hole_screw_d",
    "type": "number",
    "value": 3,
  },
  {
    "description": "The thickness of the head of the screw, so it will fit behind the hole",
    "group": "Hanging holes",
    "name": "wall_screw_head_size",
    "type": "number",
    "value": 3,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing slider.scad: slider.scad 1`] = `
[
  {
    "description": "slider widget for number with max. value",
    "group": undefined,
    "name": "sliderWithMax",
    "type": "number",
    "value": 34,
  },
  {
    "description": "slider widget for number in range",
    "group": undefined,
    "name": "sliderWithRange",
    "type": "number",
    "value": 34,
  },
  {
    "description": "step slider for number",
    "group": undefined,
    "name": "stepSlider",
    "type": "number",
    "value": 2,
  },
  {
    "description": "slider widget for number in range",
    "group": undefined,
    "name": "sliderCentered",
    "type": "number",
    "value": 0,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing specialChars.scad: specialChars.scad 1`] = `
[
  {
    "description": undefined,
    "group": "Settings",
    "name": "$fn",
    "type": "number",
    "value": 200,
  },
  {
    "description": undefined,
    "group": "Settings",
    "name": "width",
    "type": "number",
    "value": 30,
  },
  {
    "description": undefined,
    "group": "Settings",
    "name": "length",
    "type": "number",
    "value": 50,
  },
  {
    "description": undefined,
    "group": "Hidden",
    "name": "height",
    "type": "number",
    "value": 10,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing specialVector.scad: specialVector.scad 1`] = `
[
  {
    "description": "Spin box box for vector with less than or equal to 4 elements",
    "group": undefined,
    "name": "Vector2",
    "type": "string",
    "value": "[12,34]",
  },
  {
    "description": undefined,
    "group": undefined,
    "name": "Vector3",
    "type": "string",
    "value": "[12,34,45]",
  },
  {
    "description": undefined,
    "group": undefined,
    "name": "Vector4",
    "type": "string",
    "value": "[12,34,45,23]",
  },
  {
    "description": undefined,
    "group": undefined,
    "name": "VectorRange3",
    "type": "string",
    "value": "[12,34,46]",
  },
  {
    "description": undefined,
    "group": undefined,
    "name": "VectorRange4",
    "type": "string",
    "value": "[12,34,45,23]",
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing spinbox.scad: spinbox.scad 1`] = `
[
  {
    "description": "spinbox with step size 0.5",
    "group": undefined,
    "name": "Spinbox",
    "type": "number",
    "value": 5.5,
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing splitAfterFunction.scad: splitAfterFunction.scad 1`] = `
[
  {
    "description": "Text box for vector with more than 4 elements",
    "group": undefined,
    "name": "Vector6",
    "type": "string",
    "value": "[12,34,44,43,23,23]",
  },
  {
    "description": "Text box for string with length 8",
    "group": undefined,
    "name": "String",
    "type": "string",
    "value": "length",
  },
]
`;

exports[`testing parameter parsing of openscad scripts testing testbox.scad: testbox.scad 1`] = `
[
  {
    "description": "Text box for vector with more than 4 elements",
    "group": undefined,
    "name": "Vector6",
    "type": "string",
    "value": "[12,34,44,43,23,23]",
  },
  {
    "description": "Text box for string with length 8",
    "group": undefined,
    "name": "String",
    "type": "string",
    "value": "length",
  },
]
`;
